src/util/result.h

```diff
@@ -0,0 +1,109 @@
+// llvm `ErrorSuccess` or `Error` means that the bool operator is consistent.
+//
+// To indicate success, you can merely do `return OK()`.
+class OK {};
+
+// Despite the name, in LLVM `ErrorOr` corresponds to a `std::error_code`
+// whilst `Expected` corresponds to a richer LLVM `Error`. The latter is
+// broadly preferred for newer code, as it allows a much richer expression of
+// error information. In an attempt to avoid confusion, we standardize on only
+// the modern versions, but call it `Result` to make its purpose clear. In the
+// future, this could map to a newer standard, such as the standard library
+// `std::expected<T, Error>`, independently of the names in this namespace. The
+// purpose of this file is to approximately insulate against such changes (but
+// not to provide a perfect abstraction, a little bit of leakage is okay).
+//
+// For now, the LLVM `Error` class provides mandatory checking (errors cannot
+// be ignored, they must be explicitly forward or consumed), which is useful.
+//
+// You can use Result as follows:
+//
+//     Result<> foo() {
+//       return OK();
+//     }
+//
+//     Result<Value> bar() {
+//       return Value();
+//     }
+//
+template <typename T = OK>
+using Result = llvm::Expected<T>;
+
+// For error types, one can define custom error type by inheriting from
+// `ErrorInfo`. For example, you may construct an error as follows:
+//
+//    class BpfVerifierError : public ErrorInfo<BpfVerifierError> {
+//    public:
+//      static char ID;
+//      void log(llvm::raw_ostream &OS) const override {
+//        OS << "End of file reached.";
+//      }
+//    };
+//
+template <typename T>
+class ErrorInfo : public llvm::ErrorInfo<T> {
+public:
+  std::error_code convertToErrorCode() const override
+  {
+    return llvm::inconvertibleErrorCode();
+  }
+};
+
+// LLVM has its own set of stream classes and wrappers. Provide a simple
+// wrapper that supports `std::ostream` for convenience.
+std::ostream& operator<<(std::ostream& out, const llvm::Error& err);
+
+// All errors are constructed using `make_error<...>` with the error class.
+template <typename E, typename... Ts>
+auto make_error(Ts... args)
+{
+  return llvm::make_error<E>(std::forward<Ts>(args)...);
+}
+
+// For error handling there are several cases to consider:
+//
+// (1) If you want to propagate the error, you can return as expected:
+//
+//   auto ok = doAThing();
+//   if (!ok) {
+//     return ok.takeError();
+//   }
+//
+// (2) If you need to handle some cases, you can use `handleErrors`. Note that
+// you will still need to propagate unhandled cases via (1).
+//
+//   auto ok = doAThing();
+//   if (!ok) {
+//     auto nowOk = handleErrors(ok.takeError(),
+//                               [](const BpfVerifierError&) { ... });
+//     if (!nowOk) {
+//       return nowOk.takeError();
+//     }
+//  }
+template <typename Err, typename... Ts>
+Result<> handleErrors(Err err, Ts... args)
```

Comment:
Don’t you need rval ref for perfect forwarding? 
```suggestion
Result<> handleErrors(Err &&err, Ts... args)
```